crates/bevy_math/src/curves/variable.rs

```diff
@@ -0,0 +1,708 @@
+        modes.resize(2, Interpolation::Linear);
+
+        let mut tangents_control = Vec::with_capacity(2);
+        tangents_control.resize(2, TangentControl::Auto);
+
+        let mut tangents = Vec::with_capacity(2);
+        tangents.resize(2, T::FLAT_TANGENT);
+
+        if t0 < t1 {
+            Self {
+                time_stamps: vec![t0, t1],
+                keyframes: vec![v0, v1],
+                modes,
+                tangents_control,
+                tangents_in: tangents.clone(),
+                tangents_out: tangents,
+            }
+        } else {
+            Self {
+                time_stamps: vec![t1, t0],
+                keyframes: vec![v1, v0],
+                modes,
+                tangents_control,
+                tangents_in: tangents.clone(),
+                tangents_out: tangents,
+            }
+        }
+    }
+
+    pub fn from_constant(v: T) -> Self {
+        Self {
+            time_stamps: vec![0.0],
+            keyframes: vec![v],
+            modes: vec![Interpolation::Hermite],
+            tangents_control: vec![TangentControl::Auto],
+            tangents_in: vec![T::FLAT_TANGENT],
+            tangents_out: vec![T::FLAT_TANGENT],
+        }
+    }
+
+    /// Insert a new keyframe
+    ///
+    /// ```rust
+    /// use bevy_math::curves::{CurveVariable, TangentControl};
+    ///
+    /// # fn main() {
+    /// let mut curve = CurveVariable::from_constant(0.0f32);
+    /// curve.insert()
+    ///     .set_time(1.0)
+    ///     .set_value(2.0)
+    ///     .set_tangent_control(TangentControl::Flat)
+    ///     .done();
+    ///
+    /// assert_eq!(curve.len(), 2);
+    /// # }
+    /// ```
+    pub fn insert(&mut self) -> CurveVariableKeyframeBuilder<T> {
+        CurveVariableKeyframeBuilder {
+            time: self
+                .time_stamps
+                .last()
+                .copied()
+                .map_or(0.0, |t| t + 0.03333),
+            value: self.keyframes.last().unwrap().clone(),
+            mode: *self.modes.last().unwrap(),
+            tangent_control: TangentControl::Auto,
+            tangent_in: T::FLAT_TANGENT,
+            tangent_out: T::FLAT_TANGENT,
+            curve: self,
+        }
+    }
+
+    pub fn remove(&mut self, index: CurveCursor) {
+        let i = index as usize;
+
+        self.time_stamps.remove(i);
+        self.keyframes.remove(i);
+        self.modes.remove(i);
+        self.tangents_control.remove(i);
+        self.tangents_in.remove(i);
+        self.tangents_out.remove(i);
+
+        if i < self.keyframes.len() {
+            self.adjust_tangents(i);
+        }
+        if i > 0 {
+            self.adjust_tangents(i - 1);
+        }
+    }
+
+    pub fn set_value(&mut self, index: CurveCursor, value: T) {
+        let i = index as usize;
+        self.keyframes[i] = value;
+        self.adjust_tangents_with_neighbors(i);
+    }
+
+    pub fn set_time(&mut self, index: CurveCursor, time: f32) -> Option<CurveCursor> {
```

Comment:
What does this function do? Sets the keyframe selected to occur at the given time?

Reply:
And it looks like the other keyframes are adjusted so then the intervals are preserved?

Reply:
it moves a keyframe in the given index to a different point in time, then it update the tangents of the neighboring keyframes that are set to `Auto`;

Maybe I should document that

Reply:
done